### DISCUSSION TITLE
Trust-Region inner solver stopping criterion

### DISCUSSION THREAD

**Discussion by kimbergmans1** [JuliaManifolds/Manopt.jl#336] [Q&A]
Hi,

I'm using the Trust-Region solver to solve my optimization problem. I wanted to know if it was possible to set a stopping criterion for the truncated conjugate gradient.

I don't know if I can replace the inner solver, but it doesn't seem possible from the documentation.
I can't see any Stopping Criterion acting on the subsolver either.
Is there a way of doing this that I haven't found?

Thanks in advance,

Kim

**Comment by kellertuer** [comment]
Hi Kim,
that is possible with one of the newer versions (at least 0.4.41). Before that, indeed, the subsolver was kind of hard-coded into trust regions (that was the first one we had implemented with a subsolver).

But you are right, we should probably extend the documentation for trust regions here, still. But all solvers with sub solvers work the same (since 0.4.41) and there is `sub_kwargs` (see for example https://manoptjl.org/stable/solvers/augmented_Lagrangian_method/) to pass any keywords to the subsolver. Alternatively you can also pass a `sub_state=` but then you have to setup your TCG-state yourself (but can pass your own stopping criterion to that) beforehand.

It depends a bit, what you find more comfortable yourself, the `sub_kwargs` “in action” can be seen here https://manoptjl.org/stable/tutorials/HowToDebug/#Subsolver-debug, but we could maybe also provide such examples for other things than debug as well. The one thing to note here, is that the `sub_kwargs` are an array of Pairs (`=>`) where the first is a `Symbol`, so you have to pass `:stopping_criterion => ...` where you replace `...` by your stopping criterion.

So, sure I will think about extending the documentation there, since we seem to have forgotten that in 0.4.41.

I hope these ideas help, if not, maybe posting a small example would help as well

**Comment by kimbergmans1** [comment]
Hi Ronny,

Thank you for your answer.
I tried it with the unit test for the trust-region:
```
n = 10
A = Symmetric(randn(n + 1, n + 1))
M = Sphere(n)
p0 = [1.0, zeros(n)...]
f(E, p) = p' * A * p
∇f(E, p) = A * p
∇²f(M, p, X) = A * X
λ = min(eigvals(A)...)
q = trust_regions(M, f, ∇f, p0; objective_type=:Euclidean,
    sub_kwargs=[:stopping_criterion =>StopAfterIteration(3)],
    return_state=true)
```

This doesn't seems to work: 
```
* sub solver state     :
    | # Solver state for `Manopt.jl`s Truncated Conjugate Gradient Descent
    |
    | ## Parameters
    | * randomize: false
    | * trust region radius: 0.7905694150420949
    |
    | ## Stopping Criterion
    | Stop When _one_ of the following are fulfilled:
    |     Max Iteration 10:     not reached
    |     Residual reduced by factor 0.1 or power 1.0:  not reached
    |     Trust region exceeded:        not reached
    |     Curvature is negative:        not reached
    |     Model Increased:      reached
    | Overall: reached
    | This indicates convergence: No
```
After a detailed look at the trust-region and TCG source code. I see that the sub_kwargs argument is not handled, unlike in the augmented lagrangian part.

https://github.com/JuliaManifolds/Manopt.jl/blob/master/src/solvers/truncated_conjugate_gradient_descent.jl
https://github.com/JuliaManifolds/Manopt.jl/blob/master/src/solvers/trust_regions.jl

So I'm trying to use the `TruncatedConjugateGradientState `but I can't really get it to work.

Do you have any tips on how to do this?

Thanks again! :)
Kim

**Reply by kellertuer** [reply]
Sure the `TruncatedConjugateGradientState` would be a bit more “low level” to set up and pass to `sub_state`. That is really more like advanced use and one should usually use the keywords.

Concerning the keywords, it seems then we (or to be precise I) might have not been careful enough in passing that on, I will take a more careful look at that tomorrow.

**Reply by kellertuer** [reply]
I am also not sure what to provide more details on the state, basically, since you looked at the code, you have to construct the state yourself (as in the code https://github.com/JuliaManifolds/Manopt.jl/blob/38f59b0ebb20c06f90a370efde8bf6bdabc7fe16/src/solvers/trust_regions.jl#L498-L506 ) and pass that to the `sub_state` as well. Ah in those lines I see the `sub_kwargs` are missing. I will add them tomorrow together with the update of the docs then.

**Reply by kellertuer** [reply]
I #337 I adopted the few lines from ALM also to trust regions, since that was an easy fix. Then you even have `sub_stopping_criterion` directly as a shortcut compared to the approach described above.

That PR might take a short moment still, since for the next version we would also like to finish the other open PR (which depends on a PR from manifolds, which if however hopefully finished by today).

**Comment by kellertuer** [comment]
Passing keywords to the subsolver – even information like debug – is now unified in the version 0.4.45 from today.
Thanks for reporting that inconsistency.

Let me know whether that fixes your problem and whether you have further ones.
